### DISCUSSION TITLE
How can I run a function with Onclick ?

### DISCUSSION THREAD

**Discussion by Elindo586** [vercel/next.js#50930] [Help] [answered]
### Summary

For this example I am trying to run a console.log, but only after I do the Onclick button... how can I do it in Next??

Sample code below trying different approaches, with, and without the useEffect function..

1.  Why would my useEffect isn't setting the new value after I run Onclick on the first button? 

2.   Why would the console.log not run when I do OnClick on the second button? 
3. Unhandled Runtime Error
Error: Objects are not valid as a React child (found: object with keys {getText}). If you meant to render a collection of children, use an array instead.

^^^ I think, 1 and 2 work on a regular REACT, but 3 seems to be rule from REACT and doesn't seem to work in a regular REACT App either.

Bottom line of what I am looking for ----->>>  How can I get a json file, pin point the value I want, and pass it to the OnClick button and start a useEffect that would show the updated value of a useState?

Code here:

```
import { useEffect, useState } from "react";
import Image from "next/image";

import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";

import Text from "../components/text";
import db from "../utils/text.json";

export async function getStaticProps() {
  return {
    props: { db },
  };
}

export default function Home({ db }) {
  const [current1, setCurrent1] = useState("current1");
  const [current2, setCurrent2] = useState("");

  useEffect(() => {
    console.log({ current1 });
  }, [current1]);

  console.log({ current2 });

  const getText = (e) => {
    db.filter((item) => item.id === "1").map((item) => {
      return <Text key={item.id} text={item.text} />;
    });
  };

  const testing1 = { text: "testing  1" };
  const testing2 = { text: "testing 2" };

  return (
    <div className="main-index">
      <div>
        {db
          .filter((item) => item.id === "1")
          .map((item) => {
            return <Text key={item.id} text={item.text} />;
          })}
      </div>

      <Container>
        <div>
          {" "}
          Hello there {current1}
          <button
            type="button"
            name="submit"
            className="btn btn-primary"
            onClick={() => {
              setCurrent1({ getText });
            }}
          >
            Click
          </button>
        </div>{" "}
        <br />
        <div>
          Hello there {current2}
          <button
            type="button"
            name="submit"
            className="btn btn-primary"
            onClick={() => {
              setCurrent2(testing2.text);
              console.log(testing2.text);
            }}
          >
            Click
          </button>
        </div>{" "}
        <br />
      </Container>
    </div>
  );
}
```

### Example

https://github.com/Elindo586/nextprojectd/blob/main/pages/test.js

**Comment by devjiwonchoi** [comment]
Seems like you have three major issues with the `test.js`.

## - `getText` is a function, not a variable.
Since you were trying to `object destructure` the `getText` **function**,  it will cause an error no matter what you do.

## - React Child cannot be set as an Object.
As it is stated in the error message, Objects are not valid as a React child. Use an array instead.
> Error: Objects are not valid as a React child (found: object with keys {getText}). If you meant to render a collection of children, use an array instead.

To sum up, you need to set a value to `setCurrent1` by doing this:

```js
onClick={() => setCurrent1({getText})} // React Child as an Object, Function as variable ❌

onClick={() => setCurrent1([getText()])} // React Child as an Array, Function executed ✅
```

## - `return` is required in a Function Block
Your `getText` function was using a curly bracket as an arrow function but was not returning anything.
Therefore, `db.filter` did return a new filtered array, but was not returned as a value for the `getText` function.

```js
// Not returning anything. ❌
const getText = (e) => {
  db.filter((item) => item.id === "1").map((item) => {
    return <Text key={item.id} text={item.text} />;
  });
};

// Returns the filtered array. ✅
const getText = () => {
  return db
    .filter((item) => item.id === '1')
    .map((item) => <Text key={item.id} text={item.text} />)
}
```

This is the fixed result:

```js
import { useEffect, useState } from "react";
import Image from "next/image";

import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";

import Text from "../components/text";
import db from "../utils/text.json";

export async function getStaticProps() {
  return {
    props: { db },
  };
}

export default function Home({ db }) {
  const [current1, setCurrent1] = useState("current1");
  const [current2, setCurrent2] = useState("");

  useEffect(() => {
    console.log({ current1 });
  }, [current1]);

  console.log({ current2 });


  const getText = () => {
    return db  // Modified here
     .filter((item) => item.id === '1')
     .map((item) => <Text key={item.id} text={item.text} />)
  }

  const testing1 = { text: "testing  1" };
  const testing2 = { text: "testing 2" };

  return (
    <div className="main-index">
      <div>
        {db
          .filter((item) => item.id === "1")
          .map((item) => {
            return <Text key={item.id} text={item.text} />;
          })}
      </div>

      <Container>
        <div>
          {" "}
          Hello there {current1}
          <button
            type="button"
            name="submit"
            className="btn btn-primary"
            onClick={() => setCurrent1([getText()])} // Modified here
          >
            Click
          </button>
        </div>{" "}
        <br />
        <div>
          Hello there {current2}
          <button
            type="button"
            name="submit"
            className="btn btn-primary"
            onClick={() => {
              setCurrent2(testing2.text);
              console.log("hello");
            }}
          >
            Click
          </button>
        </div>{" "}
        <br />
      </Container>
    </div>
  );
}
```

**Comment by Elindo586** [comment]
Cool thanks...

So it isn't giving me the error at the first push button because the function is inside an array....  now, silly question... why is the UseEffect function not running when the first push button is changing the value of `setCurrent1`  ?  The  `Hello there {current1}`  goes to blank value after hitting the push button.

That is why I was testing push button 2 and `console.log("hello")` ....  it doesn't seem to be taking the signal from the push buttons...... but `Hello there {current2}` works just fine after hitting push button 2

**Reply by devjiwonchoi** [reply, marked as answer]
Seems like you missed out the last part of my answer, added `// Modified here` comment on the fixed result code.
> ## - `return` is required in a Function Block
> Your getText function was using a curly bracket as an arrow function but was not returning anything.
Therefore, db.filter did return a new filtered array, but was not returned as a value for the getText function.

**Reply by Elindo586** [reply]
Thank you much!
